File: R/1_Discretization/main_discretization.R
# メインファイル
# 状態変数と操作変数を離散化して2期間モデルを解く.

# environmentから変数をクリア
rm(list=ls())

source("CRRA.R")

# カリブレーション
beta  <- 0.985**30     # 割引因子
gamma <- 2.0           # 相対的危険回避度
rent  <- 1.025**30-1.0 # 純利子率

# パラメータ
nw <- 10 # 所得グリッドの数
na <- 40 # 貯蓄グリッドの数
w_max <- 1.0   # 所得グリッドの最大値
w_min <- 0.1   # 所得グリッドの最小値
a_max <- 1.0   # 貯蓄グリッドの最大値
a_min <- 0.025 # 貯蓄グリッドの最小値

# 計算開始
time_start <- proc.time()

cat('', "\n")
cat('-+- Solve two period model using discretization -+-', "\n")
cat('', "\n")

# グリッドポイントを計算
grid_w <- seq(from = w_min, to = w_max, length.out = nw)
grid_a <- seq(from = a_min, to = a_max, length.out = na)

# あらゆる(w,a)の組み合わせについて生涯効用を計算

# 初期化
obj <- matrix(data = 0.0, nrow = na, ncol = nw)

for (i in 1:nw) {
    for (j in 1:na) {
        cons   <- grid_w[i] - grid_a[j]
        if (cons > 0.0) {
            obj[j, i] <- CRRA(cons, gamma) + beta*CRRA((1.0+rent)*grid_a[j], gamma)
        } else {
            obj[j, i] <- -10000.0
        }
    }
}

# 効用を最大にする操作変数を探し出す：政策関数

# 初期化
pol <- matrix(data = 0.0, nrow = nw, ncol = 1)

# 各wについて生涯効用を最大にするaを探す
for (i in 1:nw) {
    maxl <- which.max(obj[1:na, i])
    pol[i] = grid_a[maxl]
}

# 計算時間をカウント終了
cat(" Time = ", proc.time() - time_start, "\n")
cat('', "\n")

# 図を描く

library("ggplot2")
library("extrafont")

# Figure 1: value function
df <- data.frame(grid = grid_w, policy = pol)

g1 <- ggplot()
g1 <- g1 + geom_line(data=df, aes(x = grid, y = policy), size=1.5)
g1 <- g1 + labs(title="Policy Function", x="Wage: w", y="Saving: a")
plot(g1)
ggsave(file = "Fig_pol_discr.eps", width = 8, height = 6)

File: Fortran/1_Discretization/main_discretization.f90
program main_discretization
! メインファイル:
! 状態変数と操作変数を離散化して2期間モデルを解く.

    use mod_types
    use mod_my_econ_fcn
    use mod_generate_grid

    implicit none

    ! *** カリブレーション ***
    real(dp), parameter :: beta  = 0.985**30     ! 割引因子
    real(dp), parameter :: gamma = 2.0           ! 相対的危険回避度
    real(dp), parameter :: rent  = 1.025**30-1.0 ! 純利子率

    ! *** パラメータ ***
    integer, parameter :: nw = 10 ! 所得グリッドの数
    integer, parameter :: na = 40 ! 貯蓄グリッドの数
    real(dp), parameter :: w_max = 1.0   ! 所得グリッドの最大値
    real(dp), parameter :: w_min = 0.1   ! 所得グリッドの最小値
    real(dp), parameter :: a_max = 1.0   ! 貯蓄グリッドの最大値
    real(dp), parameter :: a_min = 0.025 ! 貯蓄グリッドの最小値

    ! *** ローカル変数 ***
    integer :: i, j, maxl
    real(dp) :: cons, time_begin, time_end
    real(dp), dimension(na) :: grid_a
    real(dp), dimension(nw) :: grid_w
    real(dp), dimension(nw) :: pol
    real(dp), dimension(na, nw) :: obj
    !---------------------------------

    ! 計算時間をカウント開始
    call cpu_time( time_begin )

    write(*,*) ""
    write(*,*) " -+-+-+- Solve two period model using discretization -+-+-+-"

    ! グリッドポイントを計算
    call grid_uniform(w_min, w_max, nw, grid_w)
    call grid_uniform(a_min, a_max, na, grid_a)

    ! あらゆる(w,a)の組み合わせについて生涯効用を計算

    ! 初期化
    obj = 0.0

    do i = 1,nw
        do j = 1,na
            cons = grid_w(i) - grid_a(j)
            if (cons > 0.0) then
                obj(j, i) = CRRA(cons, gamma) + beta*CRRA((1.0+rent)*grid_a(j), gamma)
            else
                ! 消費が負値の場合、ペナルティを与えてその値が選ばれないようにする
                obj(j, i) = -10000.0;
            end if
        end do
    end do

    ! 効用を最大にする操作変数を探し出す：政策関数

    ! 初期化
    pol = 0.0

    ! 各wについて生涯効用を最大にするaを探す
    do i = 1,nw
        maxl = maxloc(obj(:,i), dim=1)
        pol(i) = grid_a(maxl)
    end do

    ! 計算時間をカウント終了
    call cpu_time( time_end )
    write (*,"(' Program finished sucessfully in', f12.9, ' seconds')") time_end - time_begin

    ! 計算結果をテキストファイルで出力
    open (10, file='policy_function.txt')
        do i = 1,nw
            write (10,*) pol(i)
        end do
    close (10)

end program main_discretization
